File: DSA/Navneetpant_590011569/Day_78/Q2.java
class Solution {

    class TrieNode {
        TrieNode[] child = new TrieNode[2];
    }

    public int findMaximumXOR(int[] nums) {

        TrieNode root = new TrieNode();


        for (int num : nums) {
            insert(root, num);
        }

        int max = 0;


        for (int num : nums) {
            max = Math.max(max, getMaxXOR(root, num));
        }

        return max;
    }
    public void insert(TrieNode root, int num) {

        TrieNode curr = root;

        for (int i = 31; i >= 0; i--) {

            int bit = (num >> i) & 1;

            if (curr.child[bit] == null) {
                curr.child[bit] = new TrieNode();
            }

            curr = curr.child[bit];
        }
    }

    public int getMaxXOR(TrieNode root, int num) {

        TrieNode curr = root;
        int result = 0;

        for (int i = 31; i >= 0; i--) {

            int bit = (num >> i) & 1;

            int opposite = 1 - bit;

            if (curr.child[opposite] != null) {
                result = result | (1 << i);
                curr = curr.child[opposite];
            } 
            else {
                curr = curr.child[bit];
            }
        }

        return result;
    }
}

File: DSA/VivaswanSingh_590023022/Day_78/q2.c
#include <stdlib.h>

typedef struct Node
{
    struct Node *child[2];
} Node;

Node* createNode()
{
    Node *node = malloc(sizeof(Node));
    node->child[0] = NULL;
    node->child[1] = NULL;
    return node;
}

void insert(Node *root, int num)
{
    Node *current = root;

    for (int bit = 30; bit >= 0; bit--)
    {
        int b = (num >> bit) & 1;

        if (current->child[b] == NULL)
            current->child[b] = createNode();

        current = current->child[b];
    }
}

int getMaxXOR(Node *root, int num)
{
    Node *current = root;
    int result = 0;

    for (int bit = 30; bit >= 0; bit--)
    {
        int b = (num >> bit) & 1;
        int opposite = 1 - b;

        if (current->child[opposite] != NULL)
        {
            result |= (1 << bit);
            current = current->child[opposite];
        }
        else
        {
            current = current->child[b];
        }
    }

    return result;
}

int findMaximumXOR(int *nums, int numsSize)
{
    Node *root = createNode();
    int answer = 0;

    insert(root, nums[0]);

    for (int i = 1; i < numsSize; i++)
    {
        int current = getMaxXOR(root, nums[i]);

        if (current > answer)
            answer = current;

        insert(root, nums[i]);
    }

    return answer;
}